app/grandchallenge/reader_studies/models.py

```diff
@@ -1508,3 +1552,31 @@ def delete(self):
 
     class Meta(RequestBase.Meta):
         unique_together = (("reader_study", "user"),)
+
+
+class AnswerWidgetKindChoices(models.TextChoices):
+    ACCEPT_REJECT = "ACCEPT_REJECT", "Accept/Reject Findings"
+
+
+class AnswerWidget(models.Model):
+
+    kind = models.CharField(
```

Comment:
This should need to be a db field, it is a property of each class and not of each instance. Best thing to do is set it to None and then set it as a class attribute in the subclass. In the base class or serialiser you can add a method to get the string value of the kind through looking it up in the text choices.

Reply:
Actually, hmm, I see now. There is a problem with the designed I proposed as the foreign key must be to a concrete model! (we could use gfk but they are slow and hard to limit to a particular set of models). Rather than having a base class and a bunch of implementations I think we will need one `Widget` class, which has a `kind` like you have done here, then the `Question` can have a `OneToOne` relationship with that. It would become much like the `Answer` model I guess.

Makes configuration more difficult, but we do not need to worry about that for now.

Reply:
@amickan did you see my second comment? Sorry!

Reply:
I'm not entirely sure I follow along. The base class I added here is not abstract and the Question model does have a OneToOne realtionship with the base class in the current implementation. What's the issue with what we have now?

Reply:
If we want to have configuration options in the future, doing it in one model like you suggest here will be messy.

Reply:
The current implementation adds an unnecessary table for the base model to the database, but that's not really an issue, is it?

Reply:
We need to link the `Question` to a `Widget` somehow. If we have one table per `WidgetType` it can only be done with a GFK. If we have one table for all `Widget`s then it can be done with a `OneToOne` relationship.

Yes, it means that the `Widget` model becomes large, but it is a proven approach that we have taken with `Answer`, `Notification` and `ComponentInterface`.

Reply:
> We need to link the `Question` to a `Widget` somehow. If we have one table per `WidgetType` it can only be done with a GFK. If we have one table for all `Widget`s then it can be done with a `OneToOne` relationship.

A concrete base class achieves that as well. The widget linked to the question is then an instance of the base class, but you can still access the subclasses methods and attributes.

Reply:
GFKs are out of the question, I agree - I remember what a nightmare they were to work with for the notifications.

Reply:
Does that definitely work? You can add a `OneToOne` relationship to `AnswerWidget` and then set the value to that of a `AcceptRejectFindingsWidget`? What happens when we have other classes of widgets? Never tried this so have to be 100% certain this is going to work first.

Reply:
> You can add a `OneToOne` relationship to `AnswerWidget` and then set the value to that of a `AcceptRejectFindingsWidget`?

Yes. 


> What happens when we have other classes of widgets?

They would all follow the pattern of the `AcceptRejectFindingsWidget`, so should be no different from working with just one subclass. 


I read this tutorial on the topic: https://realpython.com/modeling-polymorphism-django-python/#concrete-base-model 

Maybe we can quickly have a call about this, to see if I'm missing potential future scenarios that we should test with the current set-up?

Reply:
Would doing `question.widget` return an instance of the concrete class, or the base class?

If we had 10 questions, using 10 different widget concrete classes, how many extra queries would that be? I imagine that if we need the concrete classes it could need 10 joins across 10 different tables?

Reply:
`question.widget` returns an instance of the base class. You can then access the subclass with `question.widget.subclassname` and yes, that requires joining across the subclass table. That in itself is maybe not so bad, but I just noticed that in order to be able to do that, you need to know which kind the widget is and the classmethod that I added `get_kind` returns None when called on the base class... so then it does become inefficient...

Reply:
Ok, I'm glad we discussed this. I think I then agree that one single `Widget` model with a type field is the way to go after all.

Reply:
OK, I think that would be relatively common as we would need to serialize all the widgets for the reader study often. The article mentions that `django-polymorphic` addresses some of these issues but that project looks completely dead.

Reply:
Indeed, not maintained, so a bit risky to use. I will work on creating one single `Widget` model.

Reply:
I've read the article a little more, was interesting. What we do for `ComponentInterfaceValues` and `Answers` covered by their [Semi Structured Model](https://realpython.com/modeling-polymorphism-django-python/#semi-structured-model) section. I still this this is the most appropriate approach. A lot of their downsides they mention are no longer true, and the trade-offs are well understood.

Reply:
I agree